packages/jupyterlab-manager/src/manager.ts

```diff
@@ -213,14 +333,131 @@ class WidgetManager extends ManagerBase<Widget> implements IDisposable {
     return this._rendermime;
   }
 
+  /**
+   * A signal emitted when state is restored to the widget manager.
+   *
+   * #### Notes
+   * This indicates that previously-unavailable widget models might be available now.
+   */
+  get restored(): ISignal<this, void> {
+    return this._restored;
+  }
+
+  /**
+   * Whether the state has been restored yet or not.
+   */
+  get restoredStatus(): boolean {
+    return this._restoredStatus;
+  }
+
   register(data: IWidgetRegistryData) {
     this._registry.set(data.name, data.version, data.exports);
   }
 
+  /**
+   * Get a model
+   *
+   * #### Notes
+   * Unlike super.get_model(), this implementation always returns a promise and
+   * never returns undefined. The promise will reject if the model is not found.
+   */
+  async get_model(model_id: string): Promise<WidgetModel> {
+    const modelPromise = super.get_model(model_id);
+    if (modelPromise === undefined) {
+      throw new Error('widget model not found');
+    }
+    return modelPromise;
+  }
+
+  /**
+   * Register a widget model.
+   */
+  register_model(model_id: string, modelPromise: Promise<WidgetModel>): void {
+    super.register_model(model_id, modelPromise);
+
+    // Update the synchronous model map
+    modelPromise.then(model => {
+        this._modelsSync.set(model_id, model);
+        model.once('comm:close', () => {
+            this._modelsSync.delete(model_id);
+        });
+    });
+    this.setDirty();
+  }
+
+
+  /**
+   * Close all widgets and empty the widget state.
+   * @return Promise that resolves when the widget state is cleared.
+   */
+  async clear_state(): Promise<void> {
+    await super.clear_state();
+    this._modelsSync = new Map();
+    this.setDirty();
```

Comment:
Should we clear away old widgets before restoring? Actually, since we're only storing live widget models right, I think it doesn't hurt to leave old widget models in - we won't get file bloating from disconnected widgets.